File: eventmesh-sdk-go/grpc/dispatcher.go
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package grpc

import (
	"fmt"
	"github.com/apache/incubator-eventmesh/eventmesh-sdk-go/grpc/proto"
	"github.com/apache/incubator-eventmesh/eventmesh-sdk-go/log"
	"sync"
	"time"

	"github.com/panjf2000/ants"
)

var (
	// ErrTopicDispatcherExist repeated dispatcher for topic
	ErrTopicDispatcherExist = fmt.Errorf("already exist dispatcher for given topic")
)

// pooledHandler internal handler for subscribe message with goroutine pool
type pooledHandler struct {
	*ants.Pool
	handler OnMessage
	timeout time.Duration
}

// OnMessage redirect the msg with pool
func (p *pooledHandler) OnMessage(msg *proto.SimpleMessage) interface{} {
	ch := make(chan interface{})
	if err := p.Submit(func() {
		m := *msg
		ch <- p.handler(&m)
	}); err != nil {
		log.Warnf("submit msg to pool err:%v, msgID:%v", err, msg.UniqueId)
		return err
	}
	select {
	case <-time.After(p.timeout):
		log.Warnf("timeout in wait the response, msgID:%v", msg.UniqueId)
		break
	case val, ok := <-ch:
		if !ok {
			log.Warnf("wait response, msg chan closed, msgID:%v", msg.UniqueId)
		} else {
			if val != nil {
				log.Infof("reply for msg:%v", msg.UniqueId)
				return val
			}
		}
	}
	return nil
}

// messageDispatcher dispatch the message to different handler according to
// it's topic
type messageDispatcher struct {
	// topicMap key is the topic name, value is the SubscribeMessageHandler
	topicMap *sync.Map
	// poolSize concurrent for dispatch received msg
	poolsize int
	// timeout to process on message
	timeout time.Duration
}

// newMessageDispatcher create new message dispatcher
func newMessageDispatcher(ps int, tm time.Duration) *messageDispatcher {
	return &messageDispatcher{
		topicMap: new(sync.Map),
		poolsize: ps,
		timeout:  tm,
	}
}

// addHandler add msg handler
func (m *messageDispatcher) addHandler(topic string, hdl OnMessage) error {
	_, ok := m.topicMap.Load(topic)
	if ok {
		return ErrTopicDispatcherExist
	}
	pool, err := ants.NewPool(m.poolsize, ants.WithPanicHandler(func(i interface{}) {
		log.Warnf("process message failure, err:%v", i)
	}))
	if err != nil {
		return err
	}
	m.topicMap.Store(topic, &pooledHandler{
		Pool:    pool,
		handler: hdl,
		timeout: m.timeout,
	})
	return nil
}

// OnMessage dispatch the message by topic
func (m *messageDispatcher) onMessage(msg *proto.SimpleMessage) (interface{}, error) {
	// subscribe response ignore it
	if msg.Topic == "" {
		return nil, nil
	}
	val, ok := m.topicMap.Load(msg.Topic)
	if !ok {
		log.Warnf("no dispatch found for topic:%s, drop msg", msg.Topic)
		return nil, ErrTopicDispatcherExist
	}
	ph := val.(*pooledHandler)
	return ph.OnMessage(msg), nil
}

File: eventmesh-sdks/eventmesh-sdk-rust/src/transport/tcp/connection/dispatcher.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

//! Request-response correlation for the TCP transport.

use std::collections::HashMap;

use tokio::sync::{mpsc, oneshot};

use super::super::frame::Package;

pub(super) type PendingKey = (u64, String);

pub(super) enum OutboundCommand {
    Send(Package),
    Request {
        package: Package,
        key: PendingKey,
        response_tx: oneshot::Sender<Package>,
    },
}

/// Owns all pending requests. It is driven exclusively by `SocketDriver`, so
/// request correlation never requires a shared map or an asynchronous lock.
pub(super) struct RpcDispatcher {
    pending: HashMap<PendingKey, oneshot::Sender<Package>>,
}

impl RpcDispatcher {
    pub(super) fn new() -> Self {
        Self {
            pending: HashMap::new(),
        }
    }

    /// Register a request immediately before the driver writes it. A caller
    /// cancelled while its command was queued is detected without touching the
    /// socket.
    pub(super) fn register(
        &mut self,
        key: PendingKey,
        response_tx: oneshot::Sender<Package>,
    ) -> bool {
        if response_tx.is_closed() {
            return false;
        }
        self.pending.insert(key, response_tx);
        true
    }

    pub(super) fn take_response(
        &mut self,
        generation: u64,
        seq: String,
    ) -> Option<oneshot::Sender<Package>> {
        self.pending.remove(&(generation, seq))
    }

    pub(super) fn cancel(&mut self, key: &PendingKey) {
        self.pending.remove(key);
    }

    pub(super) fn connection_lost(&mut self) {
        self.pending.clear();
    }
}

/// Emits a non-blocking cancellation command when an `io()` future is
/// dropped. The dispatcher remains the only owner of the pending map.
pub(super) struct PendingCancellation {
    key: Option<PendingKey>,
    cancel_tx: mpsc::UnboundedSender<PendingKey>,
}

impl PendingCancellation {
    pub(super) fn new(key: PendingKey, cancel_tx: mpsc::UnboundedSender<PendingKey>) -> Self {
        Self {
            key: Some(key),
            cancel_tx,
        }
    }

    pub(super) fn disarm(&mut self) {
        self.key = None;
    }
}

impl Drop for PendingCancellation {
    fn drop(&mut self) {
        if let Some(key) = self.key.take() {
            let _ = self.cancel_tx.send(key);
        }
    }
}
